src/adcs_handler.c

```diff
@@ -46,242 +50,450 @@ ADCS_returnState adcs_telecommand(uint32_t * command, uint32_t length) {
+  uint8_t command[2];
+  command[0] = READ_PROGRAM_INFO_ID;
+  if (index < 0 || index > 18) {
+    return ADCS_INVALID_PARAMETERS;
+  } else {
+    command[1] = index;
+    return adcs_telecommand(command, 2);
+  }
 }
 
 /**
  * @brief
- * 		File Upload Packet
- * @param packet_number
- * 		Packet Number
- * @param file_bytes
- * 		File Bytes
+ * 		Copy program to internal flash
+ * @param index
+ * 		Table 66 - Source Program index				//* not
+ * good for services 0 : BootLoader, 1 : Internal flash program, 2 : EEPROM,
+ * 3-9: External flash program 1-7, 10-17: SD user file 1-8
+ * @attention
+ * 		flag = 0x5A overwrites the boot segment
  * @return
- * 		<if applicable, brief description of what is returned by the function>
+ * 		Success of function defined in adcs_types.h
  */
-ADCS_returnState ADCS_file_upload_packet(uint16_t packet_number, char *file_bytes){
-  uint32_t command[3];
-  command[0] = FILE_UPLOAD_PACKET_ID;
-  command[1] = packet_number;
-  command[2] = file_bytes;
-  adcs_telecommand(command, 3);
+ADCS_returnState ADCS_copy_program_internal_flash(
+    uint8_t index, uint8_t overwrite_flag) {  //* why would we need the enum
+  uint8_t command[3];
+  command[0] = COPY_PROGRAM_INTERNAL_FLASH_ID;
+  if (index < 0 || index > 18) {
+    return ADCS_INVALID_PARAMETERS;
+  } else {
+    command[1] = index;
+    //* In QB50 it's unused. Would we want to overwrite boot segment?
+    command[2] = overwrite_flag;
+    return adcs_telecommand(command, 3);
+  }
 }
 
-/**
- * @brief
- * 		Finalize Uploaded File Block
- * @param file_dest
- * 		File Destination
- * @param offset
- * 		Offset into file
- * @param block_length
- * 		Length of block
- * @return
- * 		<if applicable, brief description of what is returned by the function>
- */
-ADCS_returnState ADCS_finalize_upload_block(File_Upload_Dest file_dest, uint32_t offset, uint16_t block_length){
-  uint32_t command[4];
-  command[0] = FINALIZE_UPLOAD_BLOCK_ID;
-  command[1] = file_dest;
-  command[2] = offset;
-  command[3] = block_length;
-  adcs_telecommand(command, 4);
-}
+/*************************** BootLoader TMs ***************************/
 
 /**
  * @brief
- * 		Reset HoleMap for Upload Block
+ * 		Get BootLoader state
+ * @param uptime
+ * @param 11 boolean flags from Table 69
+ * 		sram1, sram2, sram_latch_not_recovered, sram_latch_recovered,
+ * sd_initial_err, sd_read_err, sd_write_err, external_flash_err,
+ * internal_flash_err, eeprom_err, bad_boot_reg, comms_radio_err
+ * @attention
+ * 		May be replaced with an array of flags
  * @return
- * 		<if applicable, brief description of what is returned by the function>
+ * 		Success of function defined in adcs_types.h
  */
-ADCS_returnState ADCS_reset_upload_block(){
-  uint32_t command = RESET_UPLOAD_BLOCK_ID;
-  return adcs_telecommand(&command, 1);
+ADCS_returnState ADCS_get_bootloader_state(uint16_t* uptime,
+                                           uint8_t* flags_arr) {
+  uint8_t telemetry[6];
```

Comment:
I might be missing something, but I was wondering why the frame length specified in the manual greater than the total lengths of the channels/telemetry bytes. Maybe it's not such a huge deal.

Reply:
Yeah, me neither. But that's how the frame is expected to be based on the manual. Maybe they left room for further development later.